File: android/app/src/main/java/com/lightmeter/app/filmpreview/FilmGrainModel.kt
package com.lightmeter.app.filmpreview

import kotlin.math.max

internal object FilmGrainModel {
    fun intensity(baseGrainIntensity: Double, deltaEv: Double): Double {
        require(baseGrainIntensity in 0.0..MAX_BASE_INTENSITY)
        require(deltaEv.isFinite())

        val underexposureStops = max(-deltaEv, 0.0)
        return (
            baseGrainIntensity +
                underexposureStops * UNDEREXPOSURE_GRAIN_PER_STOP
            ).coerceAtMost(MAX_GRAIN_INTENSITY)
    }

    fun baseIntensityForIso(iso: Int): Double {
        require(iso > 0)
        return (BASE_INTENSITY_AT_ISO_400 * iso / 400.0)
            .coerceIn(MIN_BASE_INTENSITY, MAX_BASE_INTENSITY)
    }

    fun noise(x: Int, y: Int): Double {
        var hash = x * 0x1f1f1f1f + y * 0x45d9f3b
        hash = hash xor (hash ushr 16)
        hash *= 0x45d9f3b
        hash = hash xor (hash ushr 16)
        return (hash and 0x00ff_ffff) / 8_388_607.5 - 1.0
    }

    const val MAX_BASE_INTENSITY = 0.1
    const val MAX_GRAIN_INTENSITY = 0.065
    private const val MIN_BASE_INTENSITY = 0.004
    private const val BASE_INTENSITY_AT_ISO_400 = 0.012
    private const val UNDEREXPOSURE_GRAIN_PER_STOP = 0.012
}

File: ios/FilmLightMeter/Domain/FilmGrainModel.swift
import Foundation

enum FilmGrainModel {
    static func intensity(
        baseGrainIntensity: Double,
        deltaEV: Double
    ) -> Double {
        precondition((0...maxBaseIntensity).contains(baseGrainIntensity))
        precondition(deltaEV.isFinite)
        return min(
            baseGrainIntensity + max(-deltaEV, 0) * underexposureGrainPerStop,
            maxGrainIntensity
        )
    }

    static func baseIntensity(forISO iso: Int) -> Double {
        precondition(iso > 0)
        return min(
            max(baseIntensityAtISO400 * Double(iso) / 400, minBaseIntensity),
            maxBaseIntensity
        )
    }

    static func noise(x: Int, y: Int) -> Double {
        var hash = x &* 0x1f1f1f1f &+ y &* 0x045d9f3b
        hash ^= hash >> 16
        hash &*= 0x045d9f3b
        hash ^= hash >> 16
        return Double(hash & 0x00ff_ffff) / 8_388_607.5 - 1
    }

    static let maxBaseIntensity = 0.1
    static let maxGrainIntensity = 0.065
    private static let minBaseIntensity = 0.004
    private static let baseIntensityAtISO400 = 0.012
    private static let underexposureGrainPerStop = 0.012
}
